### PR TITLE
Issue 2393: Support glob patterns for files

### LINKED ISSUE #2393: Allow user to use glob/wildcard in file path
Currently it is only possible to use a fully qualified path to a file or a folder, eg:
* /Users/timvw/src/github/arrow-datafusion/testing/data/wildcard/green_01.csv
* /Users/timvw/src/github/arrow-datafusion/testing/data/wildcard

In situations when a folder contains multiple "sorts" of files it would be handy to use a glob/wildcard to list relevant files, eg:
* /Users/timvw/src/github/arrow-datafusion/testing/data/wildcard/green_*.csv
(This would allow one to exclude files such as red_01.csv in that same folder)

Currently this does not seem to work due to some "incorrect/flawed" logic in local.rs / list_all / tokio::fs::metadata(&prefix).await?;

### PR DESCRIPTION
**Pull request by timvw**
# Which issue does this PR close?
Closes #2393.

# What changes are included in this PR?
When the provided prefix is not a valid path (file/directory), we will now attempt to "glob" over the prefix.

Eg: when the path is  
/testdata/blah.txt -> return /testdata/blah.txt (as before)
/testdata -> return all files that exist in /testdata (as before)
/testdata/x_*_z -> return all files that match the given pattern (assuming that there is no file/directory '/testdata/x_*_z')

### CODE REVIEW

**Comment by timvw** [comment]
Resolved failing clippy, lint and cargo fmt errors

**Review by andygrove — APPROVED** [review]
Thanks for the contribution @timvw! The changes look reasonable to me.

**Comment by tustvold** [comment]
I think the code now mixes blocking IO, as performed by glob, and non-blocking IO as performed by tokio::fs. I think we should consistently use one or the other...

**Comment by timvw** [comment]
Had a look at the implementation of tokio fs operations, and copied their (non-public) asyncify method to run the globbing on...

File: data-access/src/object_store/mod.rs [hunk 1]
```diff
@@ -91,6 +92,39 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+        }
+
+        if !is_glob_path(path) {
+            self.list_file(path).await
+        } else {
+            // take path up to first occurence of a glob char
+            let path_to_first_glob_character: Vec<&str> =
+                path.splitn(2, |c| GLOB_CHARS.contains(&c)).collect();
+            // find last occurrence of folder /
+            let path_parts: Vec<&str> = path_to_first_glob_character[0]
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:111]: I'm not intimately familiar with the Rust filesystem APIs, but relying on a '/' path separator feels like it might cause excitement on windows.

Perhaps this could use https://doc.rust-lang.org/std/path/struct.Path.html#method.parent instead?

File: data-access/src/object_store/mod.rs [hunk 2]
```diff
@@ -80,7 +81,7 @@ pub trait ObjectStore: Sync + Send + Debug {
         prefix: &str,
         suffix: &str,
     ) -> Result<FileMetaStream> {
-        let file_stream = self.list_file(prefix).await?;
+        let file_stream = self.glob_file(prefix).await?;
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:84]: It seems confusing if `list_file` doesn't resolve globs, but `list_file_with_suffix` does. Perhaps we could add a `glob_file_with_suffix`, to avoid changing the behavior of a pre-existing method? This could then do something like

```
match (is_glob_path(prefix), suffix.is_empty()) {
  (false, true) => self.list_file(prefix).await,
  (false, false) => self.list_file_with_suffix(prefix, suffix).await,
  (true, _) => { // glob logic }
}
```

`glob_file` could then just be

```
async fn glob_file(&self, prefix: &str) -> Result<FileMetaStream> {
  self.glob_file_with_suffix(prefix, "").await
}
```

**Review by tustvold — COMMENTED** [review]
Liking where this is headed :+1:

File: data-access/src/object_store/mod.rs [hunk 3]
```diff
@@ -91,6 +92,39 @@ pub trait ObjectStore: Sync + Send + Debug {
         })))
     }
 
+    /// Calls `list_file` with a glob_pattern
+    async fn glob_file(&self, path: &str) -> Result<FileMetaStream> {
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:96]:
```suggestion
    async fn glob_files(&self, path: &str) -> Result<FileMetaStream> {
```

Should probably be plural for consistency

**Comment by timvw** [comment]
Trying to find the location where to put tests to verify that globbing works:

ctx.read_csv("globbedpath", CsvReadOptions::new()) -> should return data as expected
ctx.register_csv("aggregate_test_100", "globbedpath", CsvReadOptions::new()) -> should return data as expected (when queried)

**Comment by timvw** [comment]
Afaik this should be good to go now.

**Comment by timvw** [comment]
Tests pass now on my windows machine as well :)

File: data-access/src/object_store/mod.rs [hunk 4]
```diff
@@ -102,3 +132,113 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+
+const GLOB_START_CHARS: [char; 3] = ['?', '*', '['];
+
+/// Determine whether the path contains a globbing character
+fn contains_glob_start_char(path: &str) -> bool {
+    path.chars().any(|c| GLOB_START_CHARS.contains(&c))
+}
+
+/// Filters the file_stream to only contain files that end with suffix
+async fn filter_suffix(
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:144]: I don't think this function needs to be async

**Comment by timvw** [comment]
Fixed clippy issues + documented the checks that are executed during a PR build (+ added script to execute them).

File: dev/pr-checks.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
Comment (tustvold) [inline dev/pr-checks.sh:1]: Perhaps we could update the CI to run this script, to prevent it from drifting?

File: dev/pr-checks.sh [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+set -e
+cargo fmt --all -- --check
+#cargo clippy --all-targets --workspace -- -D warnings
```
Comment (tustvold) [inline dev/pr-checks.sh:21]: ?

File: data-access/src/object_store/mod.rs [hunk 7]
```diff
@@ -102,3 +132,113 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+    fn test_longest_base_path(input: &str, expected: &str) {
+        assert_eq!(
+            find_longest_search_path_without_glob_pattern(input),
+            make_expected(input, expected),
+            "testing find_longest_search_path_without_glob_pattern with {}",
+            input
+        );
+    }
+
+    fn make_expected(input: &str, expected: &str) -> String {
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:219]: Some comments would help interpret what this is doing

File: data-access/src/object_store/mod.rs [hunk 8]
```diff
@@ -102,3 +132,113 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+        assert!(!contains_glob_start_char("/"));
+        assert!(!contains_glob_start_char("/test"));
+        assert!(!contains_glob_start_char("/test/"));
+        assert!(contains_glob_start_char("/test*"));
+        Ok(())
+    }
+
+    fn test_longest_base_path(input: &str, expected: &str) {
+        assert_eq!(
+            find_longest_search_path_without_glob_pattern(input),
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:212]: Does this mean that windows uses glob expressions with `/` as opposed to `\`?

File: data-access/src/object_store/mod.rs [hunk 9]
```diff
@@ -102,3 +132,113 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+    })))
+}
+
+fn find_longest_search_path_without_glob_pattern(glob_pattern: &str) -> String {
+    // in case the glob_pattern is not actually a glob pattern, take the entire thing
+    if !contains_glob_start_char(glob_pattern) {
+        glob_pattern.to_string()
+    } else {
+        // take all the components of the path (left-to-right) which do not contain a glob pattern
+        let components_in_glob_pattern =
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:164]: I don't think you need to collect this iterator

File: data-access/src/object_store/mod.rs [hunk 10]
```diff
@@ -102,3 +132,113 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+                        .push(component_in_glob_pattern);
+                }
+            }
+        }
+
+        let mut result = path_buf_for_longest_search_path_without_glob_pattern
+            .to_str()
+            .unwrap()
+            .to_string();
+        // when we're not at the root, append a separator
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:185]: Why is this necessary?

File: dev/pr-checks.sh (hunk 5 above)
Comment (timvw) [reply dev/pr-checks.sh]: I've renamed the script to "format-code" as that is what it's really doing.

File: dev/pr-checks.sh (hunk 6 above)
Comment (timvw) [reply dev/pr-checks.sh]: Whoops. Forgot to uncomment that line for testing each of the commands.

**Comment by tustvold** [comment]
Perhaps we could split out the formatting and CI changes into a separate PR, as I think they're someone parallel to the changes in this PR?

File: data-access/src/object_store/local.rs [hunk 11]
```diff
@@ -230,4 +230,30 @@ mod tests {
 
         Ok(())
     }
+
+    #[tokio::test]
+    async fn test_globbing() -> Result<()> {
```
Comment (tustvold) [inline data-access/src/object_store/local.rs:235]: Might want a test somewhere of globstar (i.e. **)

File: data-access/src/object_store/mod.rs [hunk 12]
```diff
@@ -102,3 +132,113 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+            Path::new(glob_pattern).components().collect::<Vec<_>>();
+        let mut path_buf_for_longest_search_path_without_glob_pattern = PathBuf::new();
+        let mut encountered_glob = false;
+        for component_in_glob_pattern in components_in_glob_pattern {
+            let component_as_str =
+                component_in_glob_pattern.as_os_str().to_str().unwrap();
+            if !encountered_glob {
+                let component_str_is_glob = contains_glob_start_char(component_as_str);
+                encountered_glob = component_str_is_glob;
+                if !encountered_glob {
```
[the remark this answers is not in the document]
Comment (timvw) [reply data-access/src/object_store/mod.rs]: Because I'm still learning how to write idiomatic rust :)

File: data-access/src/object_store/mod.rs (hunk 8 above)
Comment (timvw) [reply data-access/src/object_store/mod.rs:206]: No. The glob patterns are os agnostic and described here: https://docs.rs/glob/latest/glob/struct.Pattern.html

When we build the largest path using Path(Buf), rust prefers to use \ as path separator on windows. 
I have updated the expected values such that they are OS-independent, eg: 
&format!("{MAIN_SEPARATOR}a{MAIN_SEPARATOR}") instead of "/a/" (on unix) or "\a\a" (on windows)

File: data-access/src/object_store/local.rs (hunk 11 above)
Comment (timvw) [reply data-access/src/object_store/local.rs:235]: Done

File: data-access/src/object_store/mod.rs (hunk 7 above)
Comment (timvw) [reply data-access/src/object_store/mod.rs]: Updated the code such that this updating/fixing of expected values is not needed anymore

File: data-access/src/object_store/mod.rs (hunk 10 above)
Comment (timvw) [reply data-access/src/object_store/mod.rs:179]: Because we want to build a path as long as possible (which means including a trailing separator). 

Eg: On S3 this saves us an additional lookup 
aws s3 ls s3://bucket/foo -> PRE foo
aws s3 ls s3://bucket/foo/ -> files in foo..

Comment (timvw) [reply data-access/src/object_store/mod.rs:179]: Because we need to build a path (or key) as long as possible (which allows us to minimise the amount of list operations).

Eg:
aws s3 ls s3://bucket/foo -> PRE foo
aws s3 ls s3://bucket/foo/ -> files in foo

Comment (tustvold) [reply data-access/src/object_store/mod.rs:179]: Hmm... Fair enough, this has actually made me realise something that is a touch off in the current implementation w.r.t prefixes - will raise a ticket

File: data-access/src/object_store/mod.rs [hunk 13]
```diff
@@ -102,3 +132,118 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+
+/// Determine whether the path contains a globbing character
+fn contains_glob_start_char(path: &str) -> bool {
+    path.chars().any(|c| GLOB_START_CHARS.contains(&c))
+}
+
+/// Filters the file_stream to only contain files that end with suffix
+fn filter_suffix(file_stream: FileMetaStream, suffix: &str) -> Result<FileMetaStream> {
+    let suffix = suffix.to_owned();
+    Ok(Box::pin(file_stream.filter(move |fr| {
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:146]: try_filter would perhaps be nicer https://docs.rs/futures/latest/futures/stream/trait.TryStreamExt.html#method.try_filter

File: data-access/src/object_store/mod.rs [hunk 14]
```diff
@@ -102,3 +132,118 @@ pub trait ObjectStore: Sync + Send + Debug {
...
+        glob_pattern.to_string()
+    } else {
+        // take all the components of the path (left-to-right) which do not contain a glob pattern
+        let components_in_glob_pattern = Path::new(glob_pattern).components();
+        let mut path_buf_for_longest_search_path_without_glob_pattern = PathBuf::new();
+        for component_in_glob_pattern in components_in_glob_pattern {
+            let component_as_str =
+                component_in_glob_pattern.as_os_str().to_str().unwrap();
+            let component_str_is_glob = contains_glob_start_char(component_as_str);
+            if component_str_is_glob {
```
Comment (tustvold) [inline data-access/src/object_store/mod.rs:167]:
```suggestion
            if contains_glob_start_char(component_as_str) {
```

**Review by tustvold — APPROVED** [review]
Looks good to me, one minor nit which you can take or leave, will merge tomorrow unless anyone else ways in

File: data-access/src/object_store/mod.rs (hunk 13 above)
Comment (timvw) [reply data-access/src/object_store/mod.rs:146]: This is the code that was already in place. 

When I update the code to use try_filter, it will not pass-through the errors anymore (not sure if anyone was actually doing something with them though..)

Comment (timvw) [reply data-access/src/object_store/mod.rs:146]: Rebased to handle the merge conflict

Comment (tustvold) [reply data-access/src/object_store/mod.rs:146]: >  it will not pass-through the errors anymore

It should do? "All errors are passed through without filtering in this combinator."

File: data-access/src/object_store/mod.rs [hunk 15]
```diff
@@ -80,15 +83,42 @@ pub trait ObjectStore: Sync + Send + Debug {
         prefix: &str,
         suffix: &str,
     ) -> Result<FileMetaStream> {
-        let file_stream = self.list_file(prefix).await?;
-        let suffix = suffix.to_owned();
-        Ok(Box::pin(file_stream.filter(move |fr| {
-            let has_suffix = match fr {
-                Ok(f) => f.path().ends_with(&suffix),
-                Err(_) => true,
```
Comment (timvw) [inline data-access/src/object_store/mod.rs:88]: @tustvold -> original code where the errors were allowed to pass (Would be surprised to see anyone doing something meaningful with those..)

Comment (tustvold) [reply data-access/src/object_store/mod.rs:88]: try_filter doesn't filter out errors, see https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f7f679b078eeaf042970fc4ce212640e

So this could be simplified to something like

```
use futures::future::ready;
use futures::stream::TryStreamExt;

...

file_stream.try_filter(|f| ready(f.path().end_with(&suffix)))
```

Comment (timvw) [reply data-access/src/object_store/mod.rs:88]: Correct. 

Should have read the docs a bit better, clearly stated there: https://docs.rs/futures/0.3.1/futures/stream/trait.TryStreamExt.html#method.try_filter -> All errors are passed through without filtering in this combinator. Learning something new every day :)

Comment (timvw) [reply data-access/src/object_store/mod.rs:88]: Applied the change as you suggested. Thanks for your patience with me :)

**Comment by tustvold** [comment]
Thanks again :+1:
